src/VisualStudio/VisualBasic/Impl/Snippets/VisualBasicSnippetExpansionClientFactory.vb

```diff
@@ -0,0 +1,65 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+Imports System.Collections.Immutable
+Imports System.Composition
+Imports Microsoft.CodeAnalysis
+Imports Microsoft.CodeAnalysis.Completion
+Imports Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp
+Imports Microsoft.CodeAnalysis.Editor.[Shared].Utilities
+Imports Microsoft.CodeAnalysis.Host.Mef
+Imports Microsoft.CodeAnalysis.Options
+Imports Microsoft.VisualStudio.Editor
+Imports Microsoft.VisualStudio.LanguageServices.Implementation.Snippets
+Imports Microsoft.VisualStudio.Text
+Imports Microsoft.VisualStudio.Text.Editor
+Imports Microsoft.VisualStudio.Text.Editor.Commanding
+
+Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Snippets
+
+    <ExportLanguageService(GetType(ISnippetExpansionClientFactory), LanguageNames.VisualBasic)>
+    <[Shared]>
+    Friend NotInheritable Class VisualBasicSnippetExpansionClientFactory
+        Inherits AbstractSnippetExpansionClientFactory
+
+        Private ReadOnly _threadingContext As IThreadingContext
+        Private ReadOnly _signatureHelpControllerProvider As SignatureHelpControllerProvider
+        Private ReadOnly _editorCommandHandlerServiceFactory As IEditorCommandHandlerServiceFactory
+        Private ReadOnly _editorAdaptersFactoryService As IVsEditorAdaptersFactoryService
+        Private ReadOnly _argumentProviders As ImmutableArray(Of Lazy(Of ArgumentProvider, OrderableLanguageMetadata))
+        Private ReadOnly _editorOptionsService As EditorOptionsService
```

Comment:
nit.  *lots* of duplication between the C# and VB factories.  would love this to move down.

Reply:
➡️ The duplication does not appear to be avoidable, so for now left this one as-is.

Moving it down would add a large number of parameters to the base class, and would not simplify either of the derived type constructors.

Reply:
I think if we were going to unify this further, the trick would be to get rid of the derivations of AbstractSnippetExpansionClient, and instead move any of the different behaviors there into a language service with the appropriate methods...but one step at a time seems reasonable here.

Reply:
➡️ Now refactored in the latest commit